File: src/test/java/org/example/kalkulationsprogramm/service/KundennummerServiceTest.java
package org.example.kalkulationsprogramm.service;

import org.example.kalkulationsprogramm.domain.KundenZaehler;
import org.example.kalkulationsprogramm.repository.KundeRepository;
import org.example.kalkulationsprogramm.repository.KundenZaehlerRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;

class KundennummerServiceTest {

    private KundeRepository kundeRepository;
    private KundenZaehlerRepository kundenZaehlerRepository;
    private KundennummerService service;

    @BeforeEach
    void setUp() {
        kundeRepository = mock(KundeRepository.class);
        kundenZaehlerRepository = mock(KundenZaehlerRepository.class);
        service = new KundennummerService(kundeRepository, kundenZaehlerRepository);
    }

    @Test
    void reserviereLiefertAktuelleZaehlerNummerUndInkrementiert() {
        KundenZaehler zaehler = new KundenZaehler();
        zaehler.setId(1);
        zaehler.setNaechsteNummer(1042L);
        given(kundenZaehlerRepository.lockAndGet()).willReturn(zaehler);

        String nummer = service.reserviereNaechsteKundennummer();

        assertThat(nummer).isEqualTo("1042");
        assertThat(zaehler.getNaechsteNummer()).isEqualTo(1043L);
        verify(kundeRepository, never()).findMaxKundennummer();
    }

    @Test
    void reserviereFaelltAufMaxPlusEinsZurueckWennZaehlerFehlt() {
        given(kundenZaehlerRepository.lockAndGet()).willReturn(null);
        given(kundeRepository.findMaxKundennummer()).willReturn(Optional.of("1099"));

        String nummer = service.reserviereNaechsteKundennummer();

        assertThat(nummer).isEqualTo("1100");
    }

    @Test
    void generiereStartetBei1000WennNochKeinKunde() {
        given(kundeRepository.findMaxKundennummer()).willReturn(Optional.empty());

        assertThat(service.generiereNaechsteKundennummer()).isEqualTo("1000");
    }

    @Test
    void generiereSpringtAufFallbackBeiNichtNumerischerNummer() {
        given(kundeRepository.findMaxKundennummer()).willReturn(Optional.of("ABC"));

        assertThat(service.generiereNaechsteKundennummer()).isEqualTo("1000");
    }
}

File: src/test/kotlin/org/example/kalkulationsprogramm/service/KundennummerServiceTest.kt
package org.example.kalkulationsprogramm.service

import org.assertj.core.api.Assertions.assertThat
import org.example.kalkulationsprogramm.domain.KundenZaehler
import org.example.kalkulationsprogramm.repository.KundeRepository
import org.example.kalkulationsprogramm.repository.KundenZaehlerRepository
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.BDDMockito.given
import org.mockito.Mockito.mock
import org.mockito.Mockito.never
import org.mockito.Mockito.verify
import java.util.Optional

class KundennummerServiceTest {

    private lateinit var kundeRepository: KundeRepository
    private lateinit var kundenZaehlerRepository: KundenZaehlerRepository
    private lateinit var service: KundennummerService

    @BeforeEach
    fun setUp() {
        kundeRepository = mock(KundeRepository::class.java)
        kundenZaehlerRepository = mock(KundenZaehlerRepository::class.java)
        service = KundennummerService(kundeRepository, kundenZaehlerRepository)
    }

    @Test
    fun reserviereLiefertAktuelleZaehlerNummerUndInkrementiert() {
        val zaehler = KundenZaehler().apply {
            id = 1
            naechsteNummer = 1042L
        }
        given(kundenZaehlerRepository.lockAndGet()).willReturn(zaehler)

        val nummer = service.reserviereNaechsteKundennummer()

        assertThat(nummer).isEqualTo("1042")
        assertThat(zaehler.naechsteNummer).isEqualTo(1043L)
        verify(kundeRepository, never()).findMaxKundennummer()
    }

    @Test
    fun reserviereFaelltAufMaxPlusEinsZurueckWennZaehlerFehlt() {
        given(kundenZaehlerRepository.lockAndGet()).willReturn(null)
        given(kundeRepository.findMaxKundennummer()).willReturn(Optional.of("1099"))

        val nummer = service.reserviereNaechsteKundennummer()

        assertThat(nummer).isEqualTo("1100")
    }

    @Test
    fun generiereStartetBei1000WennNochKeinKunde() {
        given(kundeRepository.findMaxKundennummer()).willReturn(Optional.empty())

        assertThat(service.generiereNaechsteKundennummer()).isEqualTo("1000")
    }

    @Test
    fun generiereSpringtAufFallbackBeiNichtNumerischerNummer() {
        given(kundeRepository.findMaxKundennummer()).willReturn(Optional.of("ABC"))

        assertThat(service.generiereNaechsteKundennummer()).isEqualTo("1000")
    }
}
